PR - Code review - Join tables

wickathou opened the pull request:
For this PR, new tables and foreign key relations have been added to enable join functions to be run on the DB.

**Features:**
- Added new tables for owners and species
- Added foreign keys to the table animals
- New queries created using joins

Screenshots:

<a href='https://github.com/wickathou/vet-clinic-sql-practice/blob/join-tables/screenshots/day3/queries-day3-1.png' target='blank'>Task set 1<a>
  - What animals belong to Melody Pond?
  - List of all animals that are pokemon (their type is Pokemon).
  - List all owners and their animals, remember to include those that don't own any animal.
  - How many animals are there per species?
  - List all Digimon owned by Jennifer Orwell.

<a href='https://github.com/wickathou/vet-clinic-sql-practice/blob/join-tables/screenshots/day3/queries-day3-2.png' target='blank'>Task set 2<a>
  - List all animals owned by Dean Winchester that haven't tried to escape.
  - Who owns the most animals?

wickathou committed:
Added new species and owners tables
Added entries

wickathou committed:
Added entries to owners and species
Added foreign key values to the animals table

wickathou committed:
Added:
New queries using joins
Screenshots of queries made

wickathou committed:
Added new screenshots to the readme file

wickathou committed:
Fixed link to screenshots

Gambit142 commented on schema.sql:
```diff
@@ -21,3 +21,35 @@ alter table animals alter column id add generated always as identity;
 -- Day 2
 
 alter table animals add column species varchar(100) ;
+
+-- Day 3
+
+create table owners(
+id serial primary key,
+fullname varchar(100),
+age int);
+
+select * from owners;
+
+create table species(
+id serial primary key,
+name varchar(100));
+
+select * from species;
+
+alter table animals drop column species;
+alter table animals add column species_id int, add column owner_id int;
+
+select * from animals;
+
+alter table animals                                                  
+add constraint species_fk
+foreign key (species_id)
+references species(id)
+on delete cascade;
+
+alter table animals                                                 
+add constraint owners_fk
+foreign key (owner_id)
```
- [x] In the schema file, kindly remove these lines of codes:

- Line 4: `createdb vet_clinic` because I get this error

![image](https://user-images.githubusercontent.com/77106962/230142124-2c491d23-0bbc-401c-8e46-9926baf30ef7.png)

#### `createdb` is a command line utility which you can run from _**bash**_ and not from **_psql_**. To create a database from **_psql_**, use the `create database` statement like so: `create database [databasename];` 💻 

- Line 19: `alter table animals alter column id add generated always as identity;`

![image](https://user-images.githubusercontent.com/77106962/230143500-8da1998a-922b-4772-a843-ce5d281e4753.png)

#### In the `animals` table, the `id` is already set as the **_primary key_** of the table on line 15. If you want to make it serial, you can do this on line 9:  `id serial,` 😃 😎

Gambit142 commented on data.sql:
```diff
@@ -17,4 +17,51 @@ insert into animals(name,date_of_birth, weight_kg,neutered,escape_attempts) valu
+begin;
+insert into species (name) values
+vet_clinic-*# ('Pokemon'),
+vet_clinic-*# ('Digimon');
```
- [x] These lines of codes cause an error when I am inserting data into the species table. It would be nice to remove the `vet_clinic-*#` code on each line. 👍 😎 

![image](https://user-images.githubusercontent.com/77106962/230146663-9b29bec3-36a1-4ab3-9cb3-0126f9770d3a.png)

Gambit142 commented on queries.sql:
```diff
@@ -71,3 +71,143 @@ select species,min(weight_kg),max(weight_kg) from animals group by species;
+
+-- Owner's animal count 
+
+select
+owners.fullname, count(animals.name)
+from owners
+left join animals on animals.owner_id = owners.id group by owners.fullname;
+
+--     fullname     | count 
+-- -----------------+-------
+--  Dean Winchester |     2
+--  Bob             |     2
+--  Sam Smith       |     1
+--  Jodie Whittaker |     0
+--  Jennifer Orwell |     2
+--  Melody Pond     |     3
+-- (6 rows)
```
- [x] Great work done here. The query should return the owners and names of the animals they own, not the count of animals that they have. Kindly change it and write the correct query. 👍 💻

Gambit142 reviewed (changes requested):
Hi @wickathou 👋, 

Good job so far!

## Highlights
- Everything that creates or changes the table's structure is saved in the schema.sql file ✔️ 
- Everything that populates tables with data is saved in the data.sql file ✔️ 
- Everything that is a query that helps to analyze data is saved in the queries.sql file ✔️ 

There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️

- [x] The link to the screenshots that you attached in the pull request description does not display the screenshots. It shows something else as seen in the screenshot below:

![image](https://user-images.githubusercontent.com/77106962/230154075-42cca50c-c4c2-4ca3-9a50-c00d78443277.png)

#### Kindly attach the screenshots directly or fix the link. 👍 😎 

_Kindly check the comments under the review for other required changes._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

You can also consider:

N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

wickathou committed:
Fixed:
Createdb command to be ran inside psql
Left join query to find owners and the animals they own
Syntax error inside data.sql line 39 and 40
PR description links to screenshots now have the right links

Kamba56 commented on queries.sql:
```diff
@@ -71,3 +71,148 @@ select species,min(weight_kg),max(weight_kg) from animals group by species;
+
+-- Who owns the most animals
+
+select
+owners.fullname, count(animals.name)
+into table ownership_count from owners
+left join animals on animals.owner_id = owners.id group by owners.fullname;
+select * from ownership_count ;
+
+--     fullname     | count 
+-- -----------------+-------
+--  Dean Winchester |     2
+--  Bob             |     2
+--  Sam Smith       |     1
+--  Jodie Whittaker |     0
+--  Jennifer Orwell |     2
+--  Melody Pond     |     3
+-- (6 rows)
+
+select                         
+*
+from ownership_count
+where count= ( select max(count) from ownership_count);
+
+--   fullname   | count 
+-- -------------+-------
+--  Melody Pond |     3
+-- (1 row)
```
- [OPTIONAL] You can do this in one query. check the example below

```sql
SELECT owners.full_name, COUNT(animals.owners_id) FROM owners
JOIN animals ON owners.id = animals.owners_id
GROUP BY owners.full_name;
```

Kamba56 reviewed (approved):
Hi @wickathou, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏
## Highlights
- [x] Good job implementing changes requested by previous reviewer!👏👏👏
- [x] Good job using the JOIN clause to query multiple tables!👏👏👏👏👏👏

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
